### PR TITLE
Select 2 v4.0 and some other enhancements

### PR DESCRIPTION
**Pull request by mabdelfattah**

### CODE REVIEW

File: app/inputs/search_select_input.rb [hunk 1]
```diff
@@ -11,6 +11,23 @@ def input_html_options
...
+  def collection
+    if !@object.nil?
+        { get_selected_value => @object[method] }
+    else
+        { '': '' }
+    end
+  end
+
+  def get_selected_value(display_name = @options[:display_name] || "name")
+    @object.send(attributized_method_name).send(display_name) rescue ''
```
Comment (monkeyci) [inline app/inputs/search_select_input.rb]: Avoid using rescue in its modifier form.

File: app/inputs/search_select_input.rb [hunk 2]
```diff
@@ -11,6 +11,23 @@ def input_html_options
...
+      opts[:name] = super[:name] + '[]'
+    end
     super.merge opts
   end
+
+  def collection
+    if !@object.nil?
+        { get_selected_value => @object[method] }
+    else
+        { '': '' }
```
Comment (monkeyci) [inline app/inputs/search_select_input.rb]: Use 2 (not 4) spaces for indentation.

File: app/inputs/search_select_input.rb [hunk 3]
```diff
@@ -11,6 +11,23 @@ def input_html_options
...
+    if @options[:multiple]
+      opts["data-multiple"] = 'true'
+      opts[:name] = super[:name] + '[]'
+    end
     super.merge opts
   end
+
+  def collection
+    if !@object.nil?
+        { get_selected_value => @object[method] }
```
Comment (monkeyci) [inline app/inputs/search_select_input.rb]: Use 2 (not 4) spaces for indentation.

File: app/inputs/ajax_filter_input.rb [hunk 4]
```diff
@@ -26,18 +32,17 @@ def input_html_options
     opts["data-width"] = @options[:width] || "100%"
     opts["data-selected"] = get_selected_value(opts["data-display_name"])
     opts["data-order"] = @options[:order_by] if @options[:order_by]
+    opts["data-per_page"] = @options[:per_page] if @options[:per_page]
     super.merge opts
   end
 
+  def get_selected_id
+    @object.conditions.find { |c| c.attributes.map(&:name).include?(method.to_s) }.values.first.value rescue nil
```
Comment (monkeyci) [inline app/inputs/ajax_filter_input.rb]: Avoid using rescue in its modifier form.<br>Line is too long. [112/100]

File: app/inputs/ajax_filter_input.rb [hunk 5]
```diff
@@ -1,11 +1,15 @@
...
   def to_html
     input_wrapping do
       [
         label_html,
-        builder.text_field(eq_input_name, input_html_options)
+        builder.select(eq_input_name,
+                        { get_selected_value => get_selected_id },
+                        { selected: get_selected_id },
+                        input_html_options
+                      )
```
Comment (monkeyci) [inline app/inputs/ajax_filter_input.rb]: Indent ) the same as the start of the line where ( is.<br>Closing method call brace must be on the same line as the last argument when opening brace is on the same line as the first argument.

File: app/inputs/ajax_filter_input.rb [hunk 6]
```diff
@@ -1,11 +1,15 @@
...
 
   def to_html
     input_wrapping do
       [
         label_html,
-        builder.text_field(eq_input_name, input_html_options)
+        builder.select(eq_input_name,
+                        { get_selected_value => get_selected_id },
+                        { selected: get_selected_id },
+                        input_html_options
```
Comment (monkeyci) [inline app/inputs/ajax_filter_input.rb]: Use one level of indentation for parameters following the first line of a multi-line method call.

File: app/inputs/ajax_filter_input.rb [hunk 7]
```diff
@@ -1,11 +1,15 @@
...
   include ActiveAdmin::Inputs::Filters::Base
 
   def to_html
     input_wrapping do
       [
         label_html,
-        builder.text_field(eq_input_name, input_html_options)
+        builder.select(eq_input_name,
+                        { get_selected_value => get_selected_id },
+                        { selected: get_selected_id },
```
Comment (monkeyci) [inline app/inputs/ajax_filter_input.rb]: Use one level of indentation for parameters following the first line of a multi-line method call.

File: app/inputs/ajax_filter_input.rb [hunk 8]
```diff
@@ -1,11 +1,15 @@
...
+class AjaxFilterInput < Formtastic::Inputs::SelectInput
   include ActiveAdmin::Inputs::Filters::Base
 
   def to_html
     input_wrapping do
       [
         label_html,
-        builder.text_field(eq_input_name, input_html_options)
+        builder.select(eq_input_name,
+                        { get_selected_value => get_selected_id },
```
Comment (monkeyci) [inline app/inputs/ajax_filter_input.rb]: Use one level of indentation for parameters following the first line of a multi-line method call.

File: app/assets/javascripts/activeadmin_addons/select2.js [hunk 9]
```diff
@@ -184,12 +186,14 @@ $(function() {
...
-          allowClear: allowClear
+          allowClear: allowClear,
+          placeholder: ' '
         });
       } else {
         $(select).select2({
           width: DEFAULT_SELECT_WIDTH,
-          allowClear: allowClear
+          allowClear: allowClear,
+          placeholder: ' '
```
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Missing trailing comma                                           comma-dangle

File: app/assets/javascripts/activeadmin_addons/select2.js [hunk 10]
```diff
@@ -184,12 +186,14 @@ $(function() {
...
           width: 'resolve',
-          allowClear: allowClear
+          allowClear: allowClear,
+          placeholder: ' '
         });
       } else {
         $(select).select2({
           width: DEFAULT_SELECT_WIDTH,
-          allowClear: allowClear
+          allowClear: allowClear,
```
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Expected property shorthand                                      object-shorthand

File: app/assets/javascripts/activeadmin_addons/select2.js [hunk 11]
```diff
@@ -184,12 +186,14 @@ $(function() {
       if ($(select).closest('.filter_form').length > 0) {
         $(select).select2({
           width: 'resolve',
-          allowClear: allowClear
+          allowClear: allowClear,
```
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Expected property shorthand                                      object-shorthand

File: app/assets/javascripts/activeadmin_addons/select2.js [hunk 12]
```diff
@@ -163,6 +153,18 @@ $(function() {
           });
         }
 
+        var select2Config = {
+          width: width,
+          containerCssClass: 'nested-select-container',
+          minimumInputLength: minimumInputLength,
+          placeholder: ' ',
+          allowClear: true
```
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Missing trailing comma                                           comma-dangle

File: app/assets/javascripts/activeadmin_addons/select2.js [hunk 13]
```diff
@@ -163,6 +153,18 @@ $(function() {
           });
         }
 
+        var select2Config = {
+          width: width,
+          containerCssClass: 'nested-select-container',
+          minimumInputLength: minimumInputLength,
```
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Expected property shorthand                                      object-shorthand

File: app/assets/javascripts/activeadmin_addons/select2.js [hunk 14]
```diff
@@ -163,6 +153,18 @@ $(function() {
           });
         }
 
+        var select2Config = {
+          width: width,
```
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Expected property shorthand                                      object-shorthand

File: app/assets/javascripts/activeadmin_addons/select2.js [hunk 15]
```diff
@@ -163,6 +153,18 @@ $(function() {
           });
         }
 
+        var select2Config = {
```
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: All 'var' declarations must be at the top of the function scope  vars-on-top<br>Unexpected var, use let or const instead                         no-var

File: app/assets/javascripts/activeadmin_addons/select2.js [hunk 16]
```diff
@@ -96,9 +100,13 @@ $(function() {
...
             }
 
+            if (per_page) {
+              query.per_page = per_page;
+            }
+
             return query;
           },
-          results: function(data, page) {
+          processResults: function(data, params) {
```
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Expected method shorthand                                        object-shorthand<br>Missing space before function parentheses                        space-before-function-paren<br>'params' is defined but never used                               no-unused-vars

File: app/assets/javascripts/activeadmin_addons/select2.js (hunk 11 above)
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Expected property shorthand                                      object-shorthand

File: app/assets/javascripts/activeadmin_addons/select2.js (hunk 12 above)
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Missing trailing comma                                           comma-dangle

File: app/assets/javascripts/activeadmin_addons/select2.js (hunk 13 above)
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Expected property shorthand                                      object-shorthand

File: app/assets/javascripts/activeadmin_addons/select2.js (hunk 14 above)
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Expected property shorthand                                      object-shorthand

File: app/assets/javascripts/activeadmin_addons/select2.js (hunk 15 above)
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: All 'var' declarations must be at the top of the function scope  vars-on-top<br>Unexpected var, use let or const instead                         no-var

File: app/assets/javascripts/activeadmin_addons/select2.js (hunk 16 above)
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Expected method shorthand                                        object-shorthand<br>Missing space before function parentheses                        space-before-function-paren<br>'params' is defined but never used                               no-unused-vars

**Comment by tbuehl** [comment]
If you give me write access I can fix the style issues so that it can be merged @mabdelfattah

**Comment by tbuehl** [comment]
Also I would suggest in general pushing fixes to the original PR instead of opening new ones. Makes it easier to keep track of a conversation.

**Comment by mabdelfattah** [comment]
@tbuehl you have access now.

**Comment by mabdelfattah** [comment]
@tbuehl But what's wrong with the Styles?

**Comment by mabdelfattah** [comment]
@emilioeduardob What's remaining to accept this pull request?

**Comment by tbuehl** [comment]
Thank you @mabdelfattah! Hound is just complaining about a few coding style violations.

File: app/assets/javascripts/activeadmin_addons/select2.js [hunk 17]
```diff
@@ -96,9 +100,13 @@ $(function() {
               query.q[parent + '_eq'] = parentId;
             }
 
+            if (per_page) {
+              query.per_page = per_page;
```
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Identifier 'per_page' is not in camel case                       camelcase

File: app/assets/javascripts/activeadmin_addons/select2.js [hunk 18]
```diff
@@ -96,9 +100,13 @@ $(function() {
               query.q[parent + '_eq'] = parentId;
             }
 
+            if (per_page) {
```
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Identifier 'per_page' is not in camel case                       camelcase

**Comment by mabdelfattah** [comment]
@tbuehl ^

**Comment by tbuehl** [comment]
I fixed some coding style violations.

For the other violations like "Expected property shorthand object-shorthand": It could make sense to turn them off, as this feature is only available in EcmaScript 6 (http://eslint.org/docs/rules/object-shorthand) @emilioeduardob

File: app/assets/javascripts/activeadmin_addons/select2.js [hunk 19]
```diff
@@ -68,19 +70,21 @@ $(function() {
...
-          data: function(term) {
+          data: function(params) {
             var textQuery = { m: 'or' };
             fields.forEach(function(field) {
               if (field == "id") {
-                textQuery[field + '_eq'] = term;
+                textQuery[field + '_eq'] = params.term;
               } else {
-                textQuery[field + '_contains'] = term;
+                textQuery[field + '_contains'] = params.term;
```
Comment (monkeyci) [inline app/assets/javascripts/activeadmin_addons/select2.js]: Unexpected string concatenation                                  prefer-template

**Comment by emilioeduardob** [comment]
great work @mabdelfattah! I'll give it a try locally and will let you know how it goes.

**Comment by emilioeduardob** [comment]
@blackjid @ldlsegovia Is there a way to disable ES6 warning messages from the monkey?

**Comment by blackjid** [comment]
yes, there are two paths...
- disable eslint and enable jshint in the `.hound.yml` file 
- copy add a eslint config with the es6 features disabled in the root of the proyect. And add that in the `hound.yml`, algo asi

https://github.com/platanus/activeadmin_addons/blob/master/.hound.yml
```yaml
...
eslint:
  enabled: true
  config_file: ".eslintrc.json"
```
    
I would do the second one, that way if you use eslint locally you would get that configuration also...

**Comment by emilioeduardob** [comment]
Thanks @blackjid !

**Comment by emilioeduardob** [comment]
hey @mabdelfattah, I was testing locally and the `tags` and `nested` features are not working.
You can run the specs locally running `bundle exec rspec spec/features/select_2_spec.rb`. I created a gist with an updated spec for select2(is not completly updated) https://gist.github.com/emilioeduardob/aba9e0539ba9556665d8b6274e2881f3 

Running the dummy app, do the tags `numbers` work for you?

**Comment by tbuehl** [comment]
Did you find the problem @mabdelfattah?

**Comment by ldlsegovia** [comment]
I close this, because I made this functionality as part of a big refactor. Thank you anyway.
https://github.com/platanus/activeadmin_addons/tree/select2-refactor

**Comment by tbuehl** [comment]
@ldlsegovia Thanks for that! But how can I use a multi-select box with AJAX?

**Comment by tbuehl** [comment]
@ldlsegovia I found out how it works (`input_html: { multiple: 'multiple' }`). But are you aware of that the `ajax_filter` doesn't work in the refactor branch? 

```
filter :affiliation_id,
         as: :search_filter,
         minimum_input_length: 3,
         fields: [:name],
         url: proc { ajax_search_nucleus_service_providers_path }
```

This only outputs a normal input field.

**Pull request closed without merging**